### PR TITLE
3184 - Add factor api

### PR DESCRIPTION
**Pull request by sonaisrayel**

### CODE REVIEW

File: api/rest/factor.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		log.Error().Err(err).Str("body", string(c.Body())).Msg("failed to parse metadata update payload")
+		return c.SendStatus(http.StatusBadRequest)
+	}
+
+	err2, done := validateInputs(c, data)
+	if done {
+		return err2
+	}
+
+	err := updateFactor(c, data)
```
Comment (finkiel) [inline api/rest/factor.go:51]: great job

File: api/rest/factor.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		log.Error().Err(err).Str("body", string(c.Body())).Msg("Failed to update Factor table with the following")
+		return c.SendStatus(http.StatusInternalServerError)
+	}
+
+	errMessage := updateMetaData(c, data)
+	if len(errMessage) != 0 {
+		return c.Status(http.StatusBadRequest).JSON(Response{Status: "error", Message: errMessage})
+	}
+
+	return c.Status(http.StatusOK).JSON(Response{Status: "Ok", Message: "Factor table was successfully updated"})
```
Comment (finkiel) [inline api/rest/factor.go:62]: we updated 2 tables: factor and metadata_queue

File: api/rest/factor.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+// @Security ApiKeyAuth
+// @Router /factor [post]
+func FactorPostHandler(c *fiber.Ctx) error {
+	data := &FactorUpdateRequest{}
+	if err := c.BodyParser(&data); err != nil {
+		log.Error().Err(err).Str("body", string(c.Body())).Msg("failed to parse metadata update payload")
+		return c.SendStatus(http.StatusBadRequest)
+	}
+
+	err2, done := validateInputs(c, data)
```
Comment (finkiel) [inline api/rest/factor.go:46]: you don't need to declare err2 -> you can use err
in line 51 you don't need to declare new variable err:
**err = updateFactor(c, data)**
and not 
**err := updateFactor(c, data)**

File: api/rest/factor.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+// FactorPostHandler Update and enable Factor setup
+// @Description Update Factor setup (publisher is mandatory, domain is mandatory)
+// @Tags Factor
+// @Accept json
+// @Produce json
+// @Param options body FactorUpdateRequest true
+// @Success 200 {object} FactorUpdateResponse
+// @Security ApiKeyAuth
+// @Router /factor [post]
```
Comment (finkiel) [inline api/rest/factor.go:38]: this is not the path,
it's /price/factor

File: api/rest/factor.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// FactorGetHandler Get factor setup
+// @Description Get factor setup
+// @Tags Factor
+// @Accept json
+// @Produce json
+// @Param options body core.GetFactorOptions true "options"
+// @Success 200 {object} core.FactorSlice
+// @Router /factor/get [post]
```
Comment (finkiel) [inline api/rest/factor.go:148]: the path is /**price**/factor/get

File: api/rest/factor.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+func FactorGetAllHandler(c *fiber.Ctx) error {
+
+	data := &core.GetFactorOptions{}
+	if err := c.BodyParser(&data); err != nil {
+		return c.Status(http.StatusInternalServerError).JSON(Response{Status: "error", Message: "error when parsing request body"})
```
Comment (finkiel) [inline api/rest/factor.go:153]: please write in your error message which is the problematic url that we are trying to parse
something like:
error when parsing request for factor/get body

File: core/factor.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+func GetFactors(ctx context.Context, ops *GetFactorOptions) (FactorSlice, error) {
+
+	qmods := ops.Filter.QueryMod().Order(ops.Order, nil, models.FactorColumns.Publisher).AddArray(ops.Pagination.Do())
+
+	qmods = qmods.Add(qm.Select("DISTINCT *"))
+
+	mods, err := models.Factors(qmods...).All(ctx, bcdb.DB())
+	if err != nil && err != sql.ErrNoRows {
+		return nil, eris.Wrap(err, "failed to retrieve publishers")
```
Comment (finkiel) [inline core/factor.go:73]: failed to retrieve factors

File: api/rest/factor.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+// @Accept json
+// @Produce json
+// @Param options body FactorUpdateRequest true
+// @Success 200 {object} FactorUpdateResponse
+// @Security ApiKeyAuth
+// @Router /factor [post]
+func FactorPostHandler(c *fiber.Ctx) error {
+	data := &FactorUpdateRequest{}
+	if err := c.BodyParser(&data); err != nil {
+		log.Error().Err(err).Str("body", string(c.Body())).Msg("failed to parse metadata update payload")
```
[the remark this answers is not in the document]
Comment (sonaisrayel) [reply api/rest/factor.go:42]: I putted this error message
"Error when parsing factor payload"

File: api/rest/factor.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	errMessage := updateMetaData(c, data)
+	if len(errMessage) != 0 {
+		return c.Status(http.StatusBadRequest).JSON(Response{Status: "error", Message: errMessage})
+	}
+
+	return c.Status(http.StatusOK).JSON(Response{Status: "Ok", Message: "Factor table was successfully updated"})
+}
+
+func validateInputs(c *fiber.Ctx, data *FactorUpdateRequest) (error, bool) {
+	fields := []struct {
```
[the remark this answers is not in the document]
Comment (sonaisrayel) [reply api/rest/factor.go:66]: done

**Pull request merged**